src/urllib3/poolmanager.py

```diff
@@ -380,17 +383,35 @@ def urlopen(self, method, url, redirect=True, **kw):
                 if header.lower() in retries.remove_headers_on_redirect:
                     kw["headers"].pop(header, None)
 
+        def drain_and_release_conn(response):
+            try:
+                # discard any remaining response body, the connection will be
+                # released back to the pool once the entire response is read
+                response.read()
+            except (
+                HTTPError,
+                SocketError,
+                BaseSSLError,
```

Comment:
Can you please explain why we're catching less exceptions here compared to the existing `drain_and_release_explain`? https://github.com/urllib3/urllib3/blob/8c7a43b4a4ca0c8d36d55f132daa2a43d06fe3c4/src/urllib3/connectionpool.py#L763-L776

(It's probably perfectly fine, I just haven't thought about that yet.)

Reply:
Looking at exceptions.py, the exceptions TimeoutError, ProtocolError, and SSLError are subclassed from HTTPError and thus caught by HTTPError.  Since no individual action is required, using HTTPError should suffice.
I could find no code for HTTPException.  There are imports such as
`from .packages.six.moves.http_client import HTTPException  # noqa: F401`
but I could not find the code.  Does it exist?

Reply:
My coding style is not to have silent failures as they can cause insidious bugs and undesired side-effects.  Instead of "pass", would it be good practice to log a debug statement here?

Reply:
> I could find no code for HTTPException. There are imports such as
> `from .packages.six.moves.http_client import HTTPException # noqa: F401`
> but I could not find the code. Does it exist?

six is a package we use to help with Python 2/3 compatibility. `six.moves` allows to use the same name for standard library modules that were renamed. See https://six.readthedocs.io/#module-six.moves for details. And indeed, httplib was renamed to http.client in Python 3. And this is where `HTTPException` is defined: https://docs.python.org/3/library/http.client.html#http.client.HTTPException. (Unfortunately, urllib3 is deeply entangled with http.client/httplib, which is one reason why https://github.com/python-trio/hip exists.)

Why `.packages`, you may ask? Because we vendor all urllib3 dependencies, six is currently included in the source: https://github.com/urllib3/urllib3/blob/master/src/urllib3/packages/six.py

Now, you're right that if we include `HTTPException`, we don't need other classes, but that was already the case in the existing `drain_and_release_conn` function. Unless we have another reason to have two different variations of that function, we should move it to a common place where it can be called. https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

This is why I was suggesting moving that function to src/urllib3/response.py. Does that make sense?

> My coding style is not to have silent failures as they can cause insidious bugs and undesired side-effects. Instead of "pass", would it be good practice to log a debug statement here?

This makes sense in general, but in that case it's not a failure, we expect to see those exceptions in the happy path. (Maybe we'll want to refine the list in the future because some exceptions should not be silently ignored, but that's an another issue.)

Reply:
@pquentin 
I understand HTTPException and will add it to the list.  As for logging, I agree with your reasoning.

I was looking at the code base, I and I see in response.py
```
def release_conn(self):
    if not self._pool or not self._connection:
        return

    self._pool._put_conn(self._connection)
    self._connection = None
```
The function **drain_and_release_conn()** acts on the response object and not the connection.  Thus putting the function in response.py is good (DRY), but I think the function name should change.  Being in the HTTPReponse object, I would suggest a name **drain()** so the call would be:
`repoonse.draing()`

Your thoughts?